mlir-sycl/test/Transforms/constant-propagation.mlir

```diff
@@ -1,39 +1,53 @@
+
 gpu.module @kernels {
+  func.func private @init(%acc: memref<1x!sycl_accessor_1_i64_w_gb>,
+                          %ptr: memref<?xi64, 1>,
+                          %accRange: memref<?x!sycl_range_1_>,
+                          %memRange: memref<?x!sycl_range_1_>,
+                          %offset: memref<?x!sycl_id_1_>)
+
 // CHECK-LABEL:     gpu.func @k0(
-// CHECK-SAME:                   %[[VAL_0:.*]]: memref<1xi64>,
-// CHECK-SAME:                   %[[VAL_1:.*]]: i64) kernel {
-// CHECK-NEXT:        %[[VAL_2:.*]] = llvm.mlir.constant(0 : i64) : i64
-// CHECK-NEXT:        affine.store %[[VAL_2]], %[[VAL_0]][0] : memref<1xi64>
+// CHECK-SAME:                   %[[VAL_0:.*]]: memref<?xi64, 1>, %[[VAL_1:.*]]: memref<?x!sycl_range_1_>, %[[VAL_2:.*]]: memref<?x!sycl_range_1_>, %[[VAL_3:.*]]: memref<?x!sycl_id_1_>, %[[VAL_4:.*]]: i64) kernel {
+// CHECK-NEXT:        %[[VAL_5:.*]] = llvm.mlir.constant(0 : i64) : i64
+// CHECK-NEXT:        %[[VAL_6:.*]] = arith.constant 0 : index
+// CHECK-NEXT:        %[[VAL_7:.*]] = arith.constant 0 : i64
+// CHECK-NEXT:        %[[VAL_8:.*]] = memref.alloca() : memref<1x!sycl_accessor_1_i64_w_gb>
+// CHECK-NEXT:        func.call @init(%[[VAL_8]], %[[VAL_0]], %[[VAL_1]], %[[VAL_2]], %[[VAL_3]]) : (memref<1x!sycl_accessor_1_i64_w_gb>, memref<?xi64, 1>, memref<?x!sycl_range_1_>, memref<?x!sycl_range_1_>, memref<?x!sycl_id_1_>) -> ()
+// CHECK-NEXT:        %[[VAL_9:.*]] = sycl.accessor.subscript %[[VAL_8]]{{\[}}%[[VAL_7]]] : (memref<1x!sycl_accessor_1_i64_w_gb>, i64) -> memref<?xi64>
+// CHECK-NEXT:        memref.store %[[VAL_5]], %[[VAL_9]]{{\[}}%[[VAL_6]]] : memref<?xi64>
 // CHECK-NEXT:        gpu.return
 // CHECK-NEXT:      }
-  gpu.func @k0(%res: memref<1xi64>, %c: i64) kernel {
-    affine.store %c, %res[0] : memref<1xi64>
+  gpu.func @k0(%ptr: memref<?xi64, 1>,
+               %accRange: memref<?x!sycl_range_1_>,
+               %memRange: memref<?x!sycl_range_1_>,
+               %offset: memref<?x!sycl_id_1_>,
+               %c: i64) kernel {
+    %c0 = arith.constant 0 : index
+    %c0_i64 = arith.constant 0 : i64
+    %acc = memref.alloca() : memref<1x!sycl_accessor_1_i64_w_gb>
+    func.call @init(%acc, %ptr, %accRange, %memRange, %offset)
+        : (memref<1x!sycl_accessor_1_i64_w_gb>, memref<?xi64, 1>,
+           memref<?x!sycl_range_1_>, memref<?x!sycl_range_1_>,
+           memref<?x!sycl_id_1_>) -> ()
+    %res = sycl.accessor.subscript %acc[%c0_i64]
+        : (memref<1x!sycl_accessor_1_i64_w_gb>, i64) -> memref<?xi64>
+    memref.store %c, %res[%c0] : memref<?xi64>
     gpu.return
   }
 }
 
-llvm.func internal @foo(%handler: !llvm.ptr, %res: !llvm.ptr) {
+llvm.func internal @foo(%handler: !llvm.ptr, %acc: !llvm.ptr) {
   %c = llvm.mlir.constant(0 : i64) : i64
-  sycl.host.schedule_kernel %handler -> @kernels::@k0(%res, %c) : (!llvm.ptr, !llvm.ptr, i64) -> ()
+  sycl.host.schedule_kernel %handler -> @kernels::@k0(%acc: !sycl_accessor_1_i64_w_gb, %c)
```

Comment:
Question: If %c is propagated to its use within kernel @k0, can the call be modified (and the kernel modified accordingly) so %c is no longer passed as an argument ?

Reply:
That is done through a combination of a SYCL-specific pass in LLVM, which adds some property to the device binary marking this argument as unused and the SYCL RT not passing that parameter after reading that property. We cannot modify the entry point here.